File: src/com/eichinn/composite/universal/Component.java
package com.eichinn.composite.universal;

/**
 * Created by ei_chinn on 2017/5/31.
 */
public abstract class Component {
    public abstract void someOperation();

    public void addChild(Component child) {
        throw new UnsupportedOperationException("unsupported Operation");
    }

    public void removeChild(Component child) {
        throw new UnsupportedOperationException("unsupported Operation");
    }

    public Component getChildren(int index) {
        throw new UnsupportedOperationException("unsupported Operation");
    }

}

File: src/com/eichinn/visitor/extension/Component.kt
package com.eichinn.visitor.extension

/**
 * Created by ei_chinn on 2017/6/7.
 */
abstract class Component {
    val name: String

    constructor(name: String) {
        this.name = name
    }

    abstract fun someOperation()

    open fun addChild(child: Component) {
        throw UnsupportedOperationException("unsupported Operation")
    }

    open fun removeChild(child: Component) {
        throw UnsupportedOperationException("unsupported Operation")
    }

    open fun getChildren(index: Int): Component {
        throw UnsupportedOperationException("unsupported Operation")
    }

    abstract fun accept(visitor: Visitor)
}